watcher/watcher_test.go

```diff
@@ -21,6 +29,443 @@ func NewWatchHandlerMock() *WatchHandler {
+			imageIDstoWlids: map[string][]string{},
+			inputEvents: []watch.Event{
+				{
+					Type: watch.Added,
+					Object: &spdxv1beta1.SBOMSPDXv2p3{
+						ObjectMeta: v1.ObjectMeta{Name: "Unknown Image ID"},
+					},
+				},
+			},
+			expectedCommands: []apis.Command{},
+			expectedErrors:   []error{ErrUnknownImageID},
+		},
+		{
+			name: "Missing ImageID does not disrupt further processing",
+			imageIDstoWlids: map[string][]string{
+				"imageID-01": {"test-wlid-01"},
+				"imageID-02": {"test-wlid-02"},
+			},
+			inputEvents: []watch.Event{
+				{
+					Type: watch.Added,
+					Object: &spdxv1beta1.SBOMSPDXv2p3{
+						ObjectMeta: v1.ObjectMeta{Name: "Unknown Image ID"},
+					},
+				},
+				{
+					Type: watch.Added,
+					Object: &spdxv1beta1.SBOMSPDXv2p3{
+						ObjectMeta: v1.ObjectMeta{Name: "imageID-01"},
+					},
+				},
+				{
+					Type: watch.Added,
+					Object: &spdxv1beta1.SBOMSPDXv2p3{
+						ObjectMeta: v1.ObjectMeta{Name: "imageID-02"},
+					},
+				},
+			},
+			expectedCommands: []apis.Command{
+				{
+					CommandName: apis.TypeScanImages,
+					Wlid:        "test-wlid-01",
+					Args: map[string]interface{}{
+						utils.ContainerToImageIdsArg: map[string]string{},
+					},
+				},
+				{
+					CommandName: apis.TypeScanImages,
+					Wlid:        "test-wlid-02",
+					Args: map[string]interface{}{
+						utils.ContainerToImageIdsArg: map[string]string{},
+					},
+				},
+			},
+			expectedErrors: []error{ErrUnknownImageID},
+		},
+	}
+
+	for _, tc := range tt {
+		t.Run(tc.name, func(t *testing.T) {
+			k8sClient := k8sfake.NewSimpleClientset()
+
+			k8sAPI := utils.NewK8sInterfaceFake(k8sClient)
+			ksStorageClient := kssfake.NewSimpleClientset()
+			wh, _ := NewWatchHandler(context.TODO(), k8sAPI, ksStorageClient)
+
+			wh.imagesIDToWlidsMap = tc.imageIDstoWlids
+
+			commandsCh := make(chan *apis.Command)
+			errCh := make(chan error)
+
+			sbomEvents := make(chan watch.Event)
+
+			go func() {
+				for _, event := range tc.inputEvents {
+					sbomEvents <- event
+				}
+
+				close(sbomEvents)
+			}()
+
+			go wh.HandleSBOMEvents(sbomEvents, commandsCh, errCh)
+
+			actualCommands := []apis.Command{}
+			actualErrors := []error{}
+
+			var done bool
+			for !done {
+				select {
+				case command, ok := <-commandsCh:
+					if ok {
+						actualCommands = append(actualCommands, *command)
+					} else {
+						done = true
+					}
+				case err, ok := <-errCh:
+					if ok {
+						actualErrors = append(actualErrors, err)
+					} else {
+						done = true
+					}
+				}
+			}
+
+			assert.Equalf(t, tc.expectedCommands, actualCommands, "Commands should match")
+			assert.Equalf(t, tc.expectedErrors, actualErrors, "Errors should match")
+		})
+	}
+}
+
+func TestSBOMWatch(t *testing.T) {
+	t.Skipf(
+		"vladklokun: blocks and deadlocks while listening on the sbomWatcher.ResultChan(). " +
+		"Does not reproduce in a live cluster on a live Watch() object",
+	)
```

Comment:
thought: I thought this note might be useful for posterity

So that we remember that there is a test that is not passing because the test deadlocks, and if we ever get a chance, we should refactor the function to make it more testable.

I think we can fix this test if we create a `NewWatcherFunc() watch.Interface` test double, include a bi-directional channel that we own there, and then close it after the test.

Or we can employ a "shut down immediately" channel for the `SBOMWatch` function, or do some tricks with `context.Context`.